TestFoundation/TestOperationQueue.swift

```diff
@@ -168,34 +168,42 @@ class TestOperationQueue : XCTestCase {
 class AsyncOperation: Operation {
 
     private let queue = DispatchQueue(label: "async.operation.queue")
-    private let lock = NSLock()
+    private let lock = NSRecursiveLock()
```

Comment:
KVO callouts should NEVER be done inside locks btw you never know who is watching.
the lock should only be done for the critical value change of the actual assignment of the new value.

We should show good design patterns especially for long standing mysterious classes like NSOperation or NSOperationQueue. The atomicity of the values is appropriate (unfortunately we have to hit it with a sledgehammer of pthread_mutex_t instead of using the appropriate atomic operations)